File: pages/20260824170700_holding_topology_passing/analysis/julia/passing_route.jl
function destination_body(token::ThrowToken, source_body, body_count)
    token.pass || return source_body
    if token.pass_target === nothing
        body_count == 2 || throw(ArgumentError("implicit pass needs exactly two bodies"))
        return 1 - source_body
    end
    target = token.pass_target - 1
    0 <= target < body_count || throw(ArgumentError("pass target outside the body set"))
    return target
end

function destination_contact(token::ThrowToken, source_contact)
    return token.crossing ? 1 - source_contact : source_contact
end

function global_contact(body, contact)
    return body * 2 + contact
end

File: pages/20260824170700_holding_topology_passing/analysis/ruby/passing_route.rb
module PassingAnalysis
  module Route
    module_function

    def destination_body(token, source_body, body_count)
      return source_body unless token.pass
      if token.pass_target.nil?
        raise ArgumentError, "implicit pass needs exactly two bodies" unless body_count == 2

        return 1 - source_body
      end

      target = token.pass_target - 1
      raise ArgumentError, "pass target outside the body set" unless (0...body_count).cover?(target)

      target
    end

    def destination_contact(token, source_contact)
      token.crossing ? 1 - source_contact : source_contact
    end

    def global_contact(body, contact)
      body * 2 + contact
    end
  end
end
